### DISCUSSION TITLE
Zhuyin keyboard in English-Mandarin board

### DISCUSSION THREAD

**Discussion by su-aac** [UNC-Project-Open-AAC/OS-DPI#251] [Help]
I want to create a keyboard that enables typing in Chinese Zhuyin (Bopomofo) within the bilingual English-Mandarin board. However, I am having trouble displaying the Zhuyin characters on the screen. I would like to navigate from the Mandarin page to the Zhuyin keyboard page (Name: 鍵盤) using tabs. Need help!
It should look like this:
![image](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/9594cacb-24e6-42b3-8dc0-fb89d23828af)
Now look like this:
![image](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/ed4deb51-cbc6-4280-8137-22fde8e2f1a8)

File:
[English_Mandarin_grid_0605.zip](https://github.com/user-attachments/files/15594883/English_Mandarin_grid_0605.zip)

**Comment by gbishop** [comment]
In the filters for that Grid you have #role equals $role but there is not role column on that sheet. When I deleted that filter it appears to work.

**Reply by su-aac** [reply]
Thanks, Gary! It works!

**Comment by su-aac** [comment]
I need help with two problems on my English-Mandarin board:

1. The "Delete" button works on the English board but not on the Mandarin board. It doesn't remove phonemes typed with the Mandarin keyboard or words from the Mandarin homepage.
![image](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/7de3f682-2cde-46f6-9f55-b6b12e04f16a)

2. On the Mandarin keyboard "鍵盤", I can combine phonemes to make sounds (like ㄙ+ㄨ), but it doesn't show me a list of characters to choose from (like 蘇, which is my last name). I've included a video showing how I type Zhuying phonemes and expect a list of characters to appear, so I can pick the one I want. It seemed like the keyboard prediction "預測" for English and Mandarin work differently.
https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/8d9d16ad-c824-4912-9606-2c8825087d9d

Here is the file.
[Uploading English_Mandarin_grid_0627.zip…]()
[English_Mandarin_grid_0627.zip](https://github.com/user-attachments/files/16020317/English_Mandarin_grid_0627.zip)

**Comment by gbishop** [comment]
1. This happens because I use \w to match words but that only works on the
latin alphabet. Maybe I should match on non space?
2. There is no builtin composition mechanism in OS-DPI. You might be able
to approximate it using another grid like we did in the word prediction
examples.

**Reply by gbishop** [reply]
Oh, I see you are talking about the keyboard.

Your use of the non-English names in the code and spreadsheet makes it very difficult for me to help.

I'm on the keyboard looking at the 8th slot on page 2 of the autocomplete grid. It looks like: 什麼

I don't see that first character (looks a bit like an H) on the keyboard. I don't see how you can get it into the $prefix without it being there. Did I miss it?

**Comment by su-aac** [comment]
Thank you for taking the time to look into this, Gary. I wonder where the $prefix is in the spreadsheet/OS-DPI and how I can make the word be there. 

"I'm on the keyboard looking at the 8th slot on page 2 of the autocomplete grid. It looks like: 什麼"
This is correct. 什麼 means 'What' in Mandarin. For the prediction, I just searched online for the frequency rank of words and copied and pasted them to the spreadsheet of prediction "預測" (I now changed "預測" into "預測predict")

預測predict= Mandarin predict
鍵盤= Mandarin keyboard

New file:
[English_Mandarin_grid_0628.zip](https://github.com/user-attachments/files/16030826/English_Mandarin_grid_0628.zip)

**Reply by gbishop** [reply]
When you type a character on the keyboard the action updates $prefix to add the character. I don't see how you can ever trigger 什麼 without having 什 on your keyboard... The filter on your predict grid uses "starts with". To make that work to predict the two letters, you need to get the first letter into prefix somehow. 

 I know next to zero about these symbols so I may be thinking about it incorrectly.

**Reply by su-aac** [reply]
Thank you for your response, Gary. You are on point! Typing in Mandarin works differently from typing in English, and there is no character on the keyboard but the phonetic components.

I asked ChatGPT to help me explain it more thoroughly. Here's the response with my edits:
Typing in Mandarin using Zhuyin (also known as Bopomofo) involves using a phonetic input method to enter Mandarin characters. Here's how it generally works:

1. **Phonetic Input**: Zhuyin consists of 37 characters representing the sounds of Mandarin syllables. These characters correspond to initial consonants, finals (vowels and nasal finals), and tones.

2. **Typing Process**:
   - **Initials and Finals**: You type the Zhuyin characters that correspond to the initial consonant and the vowel or final sound of the Mandarin syllable you want to input. For example:
     - Typing ㄙ (represents "s") followed by ㄨ (represents "u") would form the syllable ㄙㄨ"su".
     - Typing ㄇ (represents "m") followed by ㄚ (represents "a") and a tone ˇ would form the syllableㄇㄚˇ "maˇ".

3. **Predictive Text**: After typing the phonetic components, the system predicts or suggests characters that match those components. This predictive text feature helps users quickly select the desired Mandarin characters based on the phonetic input.

4. **Selecting Characters**:  Users can then select the correct character from this list.
     - Type the syllable ㄙㄨ"su"-> auto generate a list of homophones 蘇酥甦穌(...)-> I then select the one I want 蘇.
     - Type the syllableㄇㄚˇ "maˇ"-> auto generate a list of homophones 馬瑪碼(...)-> I then select the one I want 馬.

Overall, Zhuyin input allows Mandarin speakers to input characters based on their pronunciation rather than stroke order, making it convenient for typing Mandarin on digital devices.

**Reply by gbishop** [reply]
Amazing! It's always a good day when I learn something new!

So, it seems to me you can make it work by changing your prediction spreadsheet. You want the user to type ㄙㄨ and you'll show them 蘇酥甦穌 and others. You need a column for the sequence phonetic text and a label column with the corresponding homophones. You'll filter on the first column but the grid will display the second. 

Do you understand? We can Zoom if it'll help.

**Reply by su-aac** [reply]
Wow Gary I think that will work!
Is it something like this? What column name should I put so the prediction function will work?
![image](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/cbc5a58f-34b2-46d5-b573-6d77af141635)
I think it's because I didn't do it right so it showed an error.
![image](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/46fdb4e0-e4b1-49bd-a78c-a14dcc93cb96)

File:
[English_Mandarin_grid_0701.zip](https://github.com/user-attachments/files/16056102/English_Mandarin_grid_0701.zip)

**Reply by gbishop** [reply]
Give it a simple name like "phonetics"

**Reply by su-aac** [reply]
It works when I typed "ㄙㄨ"!
![image](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/1b0ce73b-3666-4ed9-a6df-5996e06ba4f5)
But when I selected the character, the phonetics did not disapeared.
![image](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/32c85afc-12c9-4ca7-a2e7-afc111843bd7)

When I typed "ㄇㄚˇ" it won't work
![image](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/c34f8761-d952-4a38-b019-a0b0e0b78c8d)

**Comment by gbishop** [comment]
Clear the prefix when you select the word

**Reply by gbishop** [reply]
There is no magic. It knows nothing about language. You tell it to filter for rows that satisfy "starts with" $prefix and it does.

**Reply by gbishop** [reply]
The same bug that prevents your Delete from working will likely make your complete action fail similarly. I'll push a change to fix that but tomorrow morning.

**Reply by gbishop** [reply]
I've got an issue merging in the changes. This afternoon I hope to fix it.

**Reply by gbishop** [reply]
I have pushed the changes to production.

**Reply by su-aac** [reply]
Thank you for the update, Gary! I appreciate it!
The Delete function can now delete the whole word for both English and Mandarin displays. 

Is it possible to delete a single phoneme on the keyboard page and delete a word on other pages (like the "Backspace" on the laptop keyboard)? For example, on the English keyboard, if I type "cat," I can delete only the last "t," not delete the whole word "cat." Same for the Mandarin Zhuyin keyboard: if I type "ㄇㄚˇ," I can delete only the tone "ˇ" or a phoneme "ㄚ" without deleting the whole "ㄇㄚˇ."

Something like this:
https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/08cbc54a-3601-475a-9f44-c8fac65d4cae

**Reply by gbishop** [reply]
Use replace_last_letter to get a single symbol and replace_last to remove a word. You might want a separate key for those two functions. In Google Docs for example, Delete gets the last character and Control-Delete gets the last word.

**Reply by su-aac** [reply]
Thanks, Gary! 
![image](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/123387291/475fc05a-17f1-46c9-8b8c-2c83a48c7595)
I changed replace_last('') to replace_last_letter(''), which can now delete one phoneme at a time.
